src/tscore/Diags.cc

```diff
@@ -215,74 +217,20 @@ Diags::print_va(const char *debug_tag, DiagsLevel diags_level, const SourceLocat
                 va_list ap) const
 {
   ink_release_assert(diags_level < DiagsLevel_Count);
-
-  using ts::LocalBufferWriter;
-  LocalBufferWriter<1024> format_writer;
+  ts::LocalBufferWriter<1024> format_writer;
 
   // Save room for optional newline and terminating NUL bytes.
   format_writer.clip(2);
 
-  //////////////////////
-  // append timestamp //
-  //////////////////////
-  {
-    struct timeval tp = ink_gettimeofday();
-    time_t cur_clock  = (time_t)tp.tv_sec;
-    char timestamp_buf[48];
-    char *buffer = ink_ctime_r(&cur_clock, timestamp_buf);
-
-    int num_bytes_written = snprintf(&(timestamp_buf[19]), (sizeof(timestamp_buf) - 20), ".%03d", (int)(tp.tv_usec / 1000));
-
-    if (num_bytes_written > 0) {
-      format_writer.write('[');
-      format_writer.write(buffer + 4, strlen(buffer + 4));
-      format_writer.write("] ", 2);
-    }
-  }
-
-  size_t timestamp_end_offset = format_writer.size();
-
-  ///////////////////////
-  // add the thread name //
-  ///////////////////////
-  format_writer.print("{thread-name} ");
-
-  //////////////////////////////////
-  // append the diag level prefix //
-  //////////////////////////////////
-
-  format_writer.write(level_name(diags_level), strlen(level_name(diags_level)));
-  format_writer.write(": ", 2);
-
-  /////////////////////////////
-  // append location, if any //
-  /////////////////////////////
-
+  format_writer.print("[{timestamp}] ");
+  auto timestamp_offset = format_writer.size();
+  format_writer.print("{thread-name} {}: ", level_name(diags_level));
```

Comment:
I prefer printing one by one. It would make removing / reordering easier.

Reply:
I disagree. The code remains mostly one by one due to the conditional checks. Beyond that, it's easy to change the order of the arguments - because the format specifiers aren't type dependent, you don't even need to change the format string. I find changing easier in this more compact style because less code has to move.

Reply:
I don't think more compact and less code are always good. Doing multiple things in one line has the same problem as doing multiple things in one commit.

Reply:
Yes, but we don't have line by line commits, either. The real goal is coherence - a commit should be a single action. In the same way, this entire formatting is one thing and therefore should be as compact as possible. It would be, IMHO, best if this could all be done in one line in which case the output would be far more obvious, due to be able to see all the pieces in a coherent whole. E.g.
```
w.print("Found {} instances across {} bytes in {} records", count, size, n_recs);
```
is much clearer than
```
w.print("Found");
w.print("{}", instances);
// some other computations
w.print("instances across ");
ink_itoa(w.auxBuffer(), w.remaining(), size);
// etc. etc., I'm too lazy to write it all out
```
I think the former is also easier to update, again because you can see the larger goal of the effort.

Reply:
I think the first is much better. The problem is that other parts of the logic require intermediate results (such as the size of the timestamp). The point of the `OptionalAffix` wrapper is to enable the former style, by making the delimiter pairs (such as <> and []) conditional on the text, that is only output if the internal content is non-empty. A limitation of it is it just checks for `nullptr`, which it should really take a predicate (such as the call to `Location()`) to detect if the string should be written.

But this harkens back to some of the original discussion of this work, in which preference for the prefix / C style formatting was expressed (e.g., the first form). The second form is basically C++ I/O stream operators, which I think have clearly been disfavored by the community.

Reply:
My point wouldn't change even if some require the intermediate results. You can put lines as many as you want between the lines. I'm saying the second one makes clear what your are changing and what you changed.

I don't see which part looks like C++ I/O stream operators. I wrote only parameters but they all would be like:
```c
w.print("({}) ", debug_tag);
// Put 100 lines here if you want.
w.print("{}" , format_string);
```

I didn't read the discussion, but if the conclusion was put everything on one line because many people think it's better in total, then I'd have to follow it.

Reply:
There are two basic styles of formatted output, prefix and infix. The prefix style is the `printf` or C style - all of the formatting information occurs first, as a prefix, and then all of the data arguments. The infix style mixes formatting and arguments, so each argument has only its own formatting immediately preceding. The best example of infix style is C++ stream I/O, where formatting is sent to the stream, and then the data argument, then more formatting and another data argument, etc. That's the style you have here. It would be a simple text transformation to change it to C++ streaming, which is not the case with the style I prefer. E.g.
```
w << '(' << debug_tag << ')';
// put 100 lines here if you want
w << format_string;
```
The problem I have is that if you do put 100 lines there, then understanding what the intended output looks like is very difficult, as you have to put together lines that are separated by an arbitrary distance. Having it in one line makes it easy to see what the output should look like as a unit. You can see in one line there's a timestamp, then the level name, the location, the debug tag, and the format string. If I want to add or remove text, there's one place to do it, I don't have to search the code to find where that bit is added.

Reply:
> The prefix style is the printf or C style - all of the formatting information occurs first, as a prefix, and then all of the data arguments.

This is exactly what the code I wrote does. They are just separated into multiple parts - each line has one formatting information as a prefix and then the data argument. Only part that looks like C++ streaming is implicit conversion.

> Having it in one line makes it easy to see what the output should look like as a unit.

I agree but this looks the only benefit of one-line style. You ignore how diff output looks like.

> You can see in one line there's a timestamp, then the level name, the location, the debug tag, and the format string.

This is just horizontal vs vertical. I was able to read it with the comments you removed.

> If I want to add or remove text, there's one place to do it, I don't have to search the code to find where that bit is added.

That's not true. You need to find the right place ("{}") in the line instead. I think finding a block with a comment is much easier than finding a pair of curly braces.

Reply:
"each line has one formatting information as a prefix and then the data argument" - exactly like C++ I/O streams. Changing `w.print("({})", arg)` to `w << '(' << arg << ')'` is a trivial text transform, which cannot be done with the single line print example. It is basically the same as `w << setw(10) << arg` - it is infix, with the formatting for each argument directly preceding that argument.

I could also argue from consistency - if you look at the overall code base, the prefix style heavily predominates. Simply find the various `DEBUG` statements, which by the nature of `DEBUG` are all pure prefix style.

Reply:
So you think the code below is C++ style? I don't think so at all. It's just separated into multiple lines. Each line is the prefix style.
```
printf("(%s) ", debug_tag);
printf("%s ", format_string);
```

You can call it whatever, and I don't care what style it is. But you are ignoring how diff look like and this is the only thing I've been saying.

What's 3rd curly braces? How about 5th? Can you answer immediately? If so, I'm ok with it. I'll ask you.

Reply:
Your argument is really against `printf` itself. If you're going to adopt that style, why use `printf` at all instead of functions like `memcpy` or `strlcpy` or `itoa`? E.g.
```
*buff++ = '(';
memcpy(buff, debug_tag, strlen(debug_tag));
buff += strlen(debug_tag);
*buff++ = ' )';
*buff++ = ' ';
memcpy(buff, format_string, strlen(format_string));
buff += strlen(format_string);
```
That is the original form of this code, if you go back just a release or two.

Reply:
No, it doesn't. The point is total readability. I think both separate-everything (value, decoration, space, comma...) and everything-on-one-line are extreme. And it would depends on complexity of format, a number of items we want to print and other contexts. So I don't think it's 0 or 1. Do you want to print 100 items with one format sting? Do you want to separate "(%s)" into three strings? I'd say no and no.

Reply:
This discussion seems to have derailed a bit :). I think we generally agree on things, I mean, no one in their sane mind would split up say printf("%d %f %s", an_int, a_float, a_string); into multiple printf() lines. I also agree that you wouldn't make one with 100 items in it. Combining two values into a single "print" seems totally reasonable to me.

Retaining the semantics of how we would use printf() is important. As for the "ugliness" of e.g. {}, yeh, that will take some time to get used to. But, I think it's important (critical) that IF we are moving to BWprinter, we should use *only* the Python syntax (which is well documented) and our implementation should aim to be as Python compatible as possible.

Reply:
I'm just saying separating this particular line into two lines would make diff more readable when we change it in the future. It's not about format string syntax nor BFW. I would say the same thing even if it was written with printf.

I cannot draw a clear line because it depends on context as I explained above. I would not complain even if I found "name={} value={}". You may think it's inconsistent. But IMO, it's for readability consistently.

Reply:
If it's just that one case, whatever. I'll get it split.